pipe/pipeline.go

```diff
@@ -97,7 +98,51 @@ func WithDir(dir string) Option {
 // WithStdin assigns stdin to the first command in the pipeline.
 func WithStdin(stdin io.Reader) Option {
 	return func(p *Pipeline) {
-		p.stdin = stdin
+		// We don't want the first stage to close `stdin`, and it is
+		// not even necessarily an `io.ReadCloser`. So wrap it in a
+		// fake `io.ReadCloser` whose `Close()` method doesn't do
+		// anything.
+		//
+		// We could use `io.NopCloser()` for this purpose, but that
+		// would have a subtle problem. If the first stage is a
+		// `Command`, then it wants to set the `exec.Cmd`'s `Stdin` to
+		// an `io.Reader` corresponding to `p.stdin`. If `Cmd.Stdin`
+		// is an `*os.File`, then `exec.Cmd` will pass the file
+		// descriptor to the subcommand directly; there is no need to
+		// create a pipe and copy the data into the input side of the
+		// pipe. But if `p.stdin` is not an `*os.File`, then this
+		// optimization is prevented. And even worse, it also has the
+		// side effect that the goroutine that copies from `Cmd.Stdin`
+		// into the pipe doesn't terminate until that fd is closed by
+		// the writing side.
+		//
+		// That isn't always what we want. Consider, for example, the
+		// following snippet, where the subcommand's stdin is set to
+		// the stdin of the enclosing Go program, but wrapped with
+		// `io.NopCloser`:
+		//
+		//     cmd := exec.Command("ls")
+		//     cmd.Stdin = io.NopCloser(os.Stdin)
+		//     cmd.Stdout = os.Stdout
+		//     cmd.Stderr = os.Stderr
+		//     cmd.Run()
+		//
+		// In this case, we don't want the Go program to wait for
+		// `os.Stdin` to close (because `ls` isn't even trying to read
+		// from its stdin). But it does: `exec.Cmd` doesn't recognize
+		// that `Cmd.Stdin` is an `*os.File`, so it sets up a pipe and
+		// copies the data itself, and this goroutine doesn't
+		// terminate until `cmd.Stdin` (i.e., the Go program's own
+		// stdin) is closed. But if, for example, the Go program is
+		// run from an interactive shell session, that might never
+		// happen, in which case the program will fail to terminate,
+		// even after `ls` exits.
+		//
+		// So instead, in this special case, we wrap `stdin` in our
+		// own `nopCloser`, which behaves like `io.NopCloser`, except
+		// that `pipe.CommandStage` knows how to unwrap it before
+		// passing it to `exec.Cmd`.
```

Comment:
I believe that `StdinPipe()` doesn't do anything more magical than create an `os.Pipe()` and fire up `io.Copy()` or its equivalent in a goroutine. So I think that the caller could do the same, and decide itself when to close the write end of the pipe (or not close it at all, if the command doesn't care about EOF).

If the process's stdin needs to combine two sources of data (e.g., an `io.MultiReader()`), then I think that the copying of the data can't be avoided. How could the extra `os.Pipe()` be avoided in this case, even in theory?

If you have a suggestion for how things could be simplified or made more general, please let me know.

Reply:
Indeed, passing in an `os.Pipe()` should be the same as getting `Cmd.StdinPipe()`, I'm not sure if I was thinking both behaved like the `io.Pipe()` where you do need the extra goroutine. With a "real" pipe, it's passed directly into the child process so that's all fine and there isn't an extra one.

It's still annoying that you have to set this up extra instead of being able to pass in more complex types, but that remains a gotcha from the stdlib rather than this library.